### PR TITLE
[BEAM-14000] Fixes Elastic search IO doesnot work when both password and keystore are used

### PR DESCRIPTION
**Pull request by nishantjain91**
This fixes the case of elasticsearch not connecting successfully when both username/password and keystore is required

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [X] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (egalpin) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:648]: If I'm not mistaken, I _believe_ this will completely replace the instance of `HttpAsyncClientBuilder` created in `RestClientBuilder`[1].  With the previous implementation, some settings like `setSSLContext` would be overwritten but all others left left intact.  We'll need to find a way to perform this additive application of settings on the builder across both username and SSL without outright replacement of the default `HttpAsyncClientBuilder` in `RestClientBuilder`.  This can be achieved using a similar technique as what's on lines 647-667 (just below this) in this file. Ex. (not tested code)

```java
      restClientBuilder.setHttpClientConfigCallback(
          httpClientBuilder -> {
            if (getUsername() != null) {
              final CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
              credentialsProvider.setCredentials(
                  AuthScope.ANY, new UsernamePasswordCredentials(getUsername(), getPassword()));
              httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
            }
            if (getKeystorePath() != null && !getKeystorePath().isEmpty()) {
              KeyStore keyStore = null;
              keyStore = KeyStore.getInstance("jks");
              try (InputStream is = new FileInputStream(new File(getKeystorePath()))) {
                String keystorePassword = getKeystorePassword();
                keyStore.load(is, (keystorePassword == null) ? null : keystorePassword.toCharArray());
              }
              final TrustStrategy trustStrategy =
                  isTrustSelfSignedCerts() ? new TrustSelfSignedStrategy() : null;
              final SSLContext sslContext =
                  SSLContexts.custom().loadTrustMaterial(keyStore, trustStrategy).build();
              final SSLIOSessionStrategy sessionStrategy = new SSLIOSessionStrategy(sslContext);
              httpClientBuilder.setSSLContext(sslContext).setSSLStrategy(sessionStrategy);
            }
            return httpClientBuilder;
          });
```

[1] https://github.com/elastic/elasticsearch/blob/v7.8.0/client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java#L209-L213

**Review by egalpin — CHANGES_REQUESTED** [review]
Thanks @nishantjain91 for getting this started!  Are there tests that exercise these code paths?  If not we should at least verify the happy path.

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java
Comment (nishantjain91) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: Made the changes requireed for this.

**Comment by nishantjain91** [comment]
For test, can you guide me how to setup a elastic cluster with ssl and username/password configuration

**Comment by egalpin** [comment]
@nishantjain91 I'm thinking it might be sufficient to test that calling `ConnectionConfiguration#createClient` with username/password, ssl, and then both set, would generate a RestClient with the expected settings configured.  That way we'll be testing the code that we have control over and not SSL/Elasticsearch's configuration of SSL.

**Comment by nishantjain91** [comment]
@egalpin I added the unit test case. 
At the moment, I added the test case in elasticsearch 7 client, but I think that might not be appropriate place for this unit test as it does not need to startup a elasticsearch server in docker to test.

It would be possible to move this test in same package as elasticsearch but not sure whether that would be correct or not.

COuld you please review it again and let me know if more changes are required

**Comment by egalpin** [comment]
@nishantjain91 Overall this change looks good at the moment, thanks! The blocker now is that we need to determine the best structure for test suites. 

I think the best approach for now would be to include a test in each of the suites for each of the supported ES versions. The ES tests could do with restructuring overall, but I don’t think your PR should be blocked by that additional effort. 

So please add a case for this test in the suite of each supported version and then I feel we can merge. You may need to rebase on master as support for version 8.x of ES was recently added (which simultaneously dropped support for version 2.x of ES).

**Comment by egalpin** [comment]
Run Java PreCommit

**Comment by nishantjain91** [comment]
@egalpin  Made the changes.

But java pre-commit is failing in pulsar IO which is not related to any changes made in this pr

File: sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java
Comment (egalpin) [inline sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java:1243]: can you add assertions to verify that the `setHttpClientConfigCallback` is as expected? I think that assertion is one that would have failed prior to your changes, and will pass now i.e. validates that your change fixes a preexisting issue.  After that this looks good to merge!

**Review by egalpin — CHANGES_REQUESTED** [review]

**Comment by aaltay** [comment]
What is the next step on this PR?

**Comment by egalpin** [comment]
This PR has been blocked on me, apologies for the slow review cadence.  The next step would be to ideally find a way to write a test that "would have failed" with the code in master now, and would now pass as a result of @nishantjain91's great changes.  That way we can guard against regressions in the future.

I haven't had time to dig into what options there might be for inspecting the resulting `RestClient` object (or the effect of the fix on generated requests via the `RestClient`, which might also be a sufficient test).

**Comment by egalpin** [comment]
@nishantjain91 have you had time to investigate possible ways to test as well?  Apologies for not having had the time yet myself.

**Comment by nishantjain91** [comment]
@egalpin I tried to use `perfromrequest` method but the problem is again same - response does not contain any useful information. 

So the two options I see stillare same - reflection or cluster with appropriate settings(.

**Comment by egalpin** [comment]
Thanks @nishantjain91 👍  @aaltay I'm actually stuck philosophically regarding how to go about testing and what ought to be tested.  I don't expect that there should be tests across the codebase to ensure the correct usage of other APIs and external modules nor their behaviour.  Ex. it would be silly to test the built-in Java addition operator.

In the case covered by @nishantjain91's work, I suppose the most correct way to test is by integration test against an ES cluster with certs.  That just strikes me as a lot of overhead for one use case.  That said, if it happened to be the most common use case, it would be well worth the overhead.

Thoughts?

**Comment by aaltay** [comment]
> Thanks @nishantjain91 👍 @aaltay I'm actually stuck philosophically regarding how to go about testing and what ought to be tested. I don't expect that there should be tests across the codebase to ensure the correct usage of other APIs and external modules nor their behaviour. Ex. it would be silly to test the built-in Java addition operator.
> 
> In the case covered by @nishantjain91's work, I suppose the most correct way to test is by integration test against an ES cluster with certs. That just strikes me as a lot of overhead for one use case. That said, if it happened to be the most common use case, it would be well worth the overhead.
> 
> Thoughts?

Integration tests for ES would probably be a reasonable option.

**Comment by egalpin** [comment]
Thanks @aaltay 👍   I'd propose that we move forward with merging this code as-is despite lack of test coverage at the moment, and separately track adding test coverage[1].  Though I don't like that idea in general, the existing code has no test coverage either (hence the need for this change now rather than having caught the issue previously), and the changes in this PR represent a net-positive change in functionality without regression of existing tests.

@nishantjain91 can you confirm whether username + password is used in any of the tests?

Thoughts?

[1] https://issues.apache.org/jira/browse/BEAM-14531

**Comment by aaltay** [comment]
What is the next steps now? It looks like there is a change request, was it addressed?

**Comment by egalpin** [comment]
IMO the next step is to merge this in.  We have https://issues.apache.org/jira/browse/BEAM-14531 (migrated to https://github.com/apache/beam/issues/21636) to track improved testing, but this change as-is represents a net improvement.

**Review by egalpin — APPROVED** [review]
Thanks @nishantjain91, and sincere apologies for such a drawn-out review process!

**Pull request merged**
